datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/recovery/SdxUpgradeRecoveryService.java

```diff
@@ -0,0 +1,107 @@
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.dto.NameOrCrn;
+import com.sequenceiq.cloudbreak.auth.altus.EntitlementService;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.event.ResourceEvent;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.cloudbreak.message.CloudbreakMessagesService;
+import com.sequenceiq.datalake.entity.DatalakeStatusEnum;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.entity.SdxStatusEntity;
+import com.sequenceiq.datalake.flow.SdxReactorFlowManager;
+import com.sequenceiq.datalake.service.sdx.SdxService;
+import com.sequenceiq.datalake.service.sdx.status.SdxStatusService;
+import com.sequenceiq.flow.api.model.FlowIdentifier;
+import com.sequenceiq.sdx.api.model.SdxRecoveryRequest;
+import com.sequenceiq.sdx.api.model.SdxRecoveryResponse;
+import com.sequenceiq.sdx.api.model.SdxRecoveryType;
+
+@Component
+public class SdxUpgradeRecoveryService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxUpgradeRecoveryService.class);
+
+    private static final Set<DatalakeStatusEnum> RECOVERABLE_STATUSES =
+            new LinkedHashSet<>(Arrays.asList(DATALAKE_UPGRADE_FAILED,
+                    RUNNING));
+
+    @Inject
+    private SdxService sdxService;
+
+    @Inject
+    private SdxStatusService sdxStatusService;
+
+    @Inject
+    private SdxReactorFlowManager sdxReactorFlowManager;
+
+    @Inject
+    private CloudbreakMessagesService messagesService;
+
+    @Inject
+    private EntitlementService entitlementService;
+
+    public SdxRecoveryResponse triggerRecovery(String userCrn, NameOrCrn clusterNameOrCrn, SdxRecoveryRequest recoverRequest) {
+        SdxCluster cluster;
+        if (clusterNameOrCrn.hasName()) {
+            cluster = sdxService.getByNameInAccount(userCrn, clusterNameOrCrn.getName());
+        } else {
+            cluster = sdxService.getByCrn(userCrn, clusterNameOrCrn.getCrn());
+        }
+        return initSdxRecovery(userCrn, recoverRequest, cluster);
+    }
+
+    private SdxRecoveryResponse initSdxRecovery(String userCrn, SdxRecoveryRequest request, SdxCluster cluster) {
+
+        // Request time validations here
+        validateDatalakeStatus(cluster);
+        // Fetch latest backup id for given runtime from datalake-dr service
+        String backupId = "";
+
+        FlowIdentifier flowIdentifier = triggerDatalakeUpgradeRecoveryFlow(request.getType(), cluster);
+        String message = getMessage(backupId);
+        return new SdxRecoveryResponse(message, flowIdentifier);
+    }
+
+    private void validateDatalakeStatus(SdxCluster cluster) {
```

Comment:
I think we should validate that there was at least a `DATALAKE_UPGRADE_FAILED` status before, if the latest status is `RUNNING`

Reply:
`RUNNING` is only there temporarily, to be able to test without actual failed upgrades, I am externalising the values to be able to remove `RUNNING` once testing has passed.